manual/core/query_timestamps/README.md

```diff
@@ -0,0 +1,137 @@
+
+### Timestamp generator
+
+The driver has a timestamp generator that gets invoked for every outgoing request; it either assigns
+a client-side timestamp to the request, or indicates that the server should assign it. 
+
+The timestamp generator is defined in the [configuration](../configuration/).
+
+#### `AtomicTimestampGenerator`
+
+```
+datastax-java-driver.request.timestamp-generator {
+  class = com.datastax.oss.driver.api.core.time.AtomicTimestampGenerator
+}
+```
+
+This is the default implementation. It always generates a client timestamp, and guarantees
+monotonicity (i.e. ever-increasing timestamps) across all application threads.
+
+Note that, in order to achieve monotonicity, the generator might return timestamps that drift out in
+the future. This happens if timestamps are generated at a rate of more than one per microsecond, or
+more likely in the event of a system clock skew. When this happens, the generator logs a warning
+message in the category `com.datastax.oss.driver.api.core.time.MonotonicTimestampGenerator`:
+
+```
+Clock skew detected: current tick (...) was ... microseconds behind the last generated timestamp (...),
+returned timestamps will be artificially incremented to guarantee monotonicity.
+```
+
+You can control that message with those options:
+
+```
+datastax-java-driver.request.timestamp-generator {
+  drift-warning {
+    # How far in the future timestamps are allowed to drift before the warning is logged.
+    # If it is undefined or set to 0, warnings are disabled.
+    threshold = 1 second
+    # How often the warning will be logged if timestamps keep drifting above the threshold.
+    interval = 10 seconds
+  }
+}
+```
+
+The generator strives to achieve microsecond resolution on a best-effort basis. But in practice, the
+real accuracy of generated timestamps is largely dependent on the granularity of the operating
+system's clock. For most systems, this minimum granularity is millisecond, and the sub-millisecond
+part is simply a counter that gets incremented until the next clock tick, as provided by
+`System.currentTimeMillis()`.
+                                                                                          
+On some systems, however, it is possible to have a better granularity by using a [JNR] call to
+[gettimeofday]. This native call will be used when available, unless use of the Java clock is forced
+with this configuration option: 
+
+```
+datastax-java-driver.request.timestamp-generator {
+  force-java-clock = true
+}
+```
+
+To check what the driver is currently using, turn on `INFO` logs for the category
+`com.datastax.oss.driver.internal.core.time`, and look for one of the following messages at
+initialization:
+
+* `Using Java system clock because this was explicitly required in the configuration`
+* `Could not access native clock (see debug logs for details), falling back to Java system clock`
+* `Using native clock for microsecond precision`
+
+#### `ThreadLocalTimestampGenerator`
```

Comment:
I wonder how useful `ThreadLocalTimestampGenerator` is.  Do we expect it to be that much of a performance improvement?  I can see it doing more harm than good, so maybe it's worth getting rid of but I don't feel too strongly about it.   Also we should make it clear which thread the timestamp is generated on the calling thread.